polkadot/node/core/candidate-validation/src/lib.rs

```diff
@@ -760,7 +755,7 @@ trait ValidationBackend {
 					PossiblyInvalidError::AmbiguousWorkerDeath |
 					PossiblyInvalidError::AmbiguousJobDeath(_),
 				)) if num_death_retries_left > 0 => num_death_retries_left -= 1,
-				Err(ValidationError::Invalid(WasmInvalidCandidate::JobError(_)))
+				Err(ValidationError::PossiblyInvalid(PossiblyInvalidError::JobError(_)))
 					if num_job_error_retries_left > 0 =>
 					num_job_error_retries_left -= 1,
 				Err(ValidationError::Internal(_)) if num_internal_retries_left > 0 =>
```

Comment:
Sorry to be pedantic, again but maybe instead of `_` below we can enumerate all the possible variants, since there shouldn't be many left: just `Prepare` and `Invalid`. This makes it more explicit, and also ensures that we never miss handling `PossiblyInvalid`. Actually, this comment isn't even pedantic but genuinely useful.

Reply:
That's actually a nice catch!

Reply:
Hmm, you're right. I'd still refactor it to move the ifs out of the guards. It's worth enumerating all the variants here.

Reply:
By moving the ifs into the branch, every branch would become `if ... else break`. Probably not worth doing. Another it-ain't-broke-don't-fix-it example lol.

Reply:
I know, but it'd be really nice to enumerate all the variants here. Some extra verbosity seems worth it to maintain correctness.

Reply:
Are you really sure about this?

```rs
match validation_result {
	Err(ValidationError::PossiblyInvalid(
		PossiblyInvalidError::AmbiguousWorkerDeath |
		PossiblyInvalidError::AmbiguousJobDeath(_),
	)) =>
		if num_death_retries_left > 0 {
			num_death_retries_left -= 1;
		} else {
			break;
		},

	Err(ValidationError::PossiblyInvalid(PossiblyInvalidError::JobError(_))) =>
		if num_job_error_retries_left > 0 {
			num_job_error_retries_left -= 1;
		} else {
			break;
		},

	Err(ValidationError::Internal(_)) =>
		if num_internal_retries_left > 0 {
			num_internal_retries_left -= 1;
		} else {
			break;
		},

	Ok(_) | Err(ValidationError::Invalid(_) | ValidationError::Preparation(_)) => break,
}
```

Reply:
Hmm, I don't think it's *too* bad. I'm worried that someone in the future may add a `PossiblyInvalid` case and forget to update this spot, because the `_` swallows it.

Maybe a small helper macro would be less verbose and more clear, e.g.

```rs
	Err(ValidationError::Internal(_)) => retry_or_break!(num_internal_retries_left),
```

Or would it just obscure this code more? cc @s0me0ne-unkn0wn for a second opinion.

Reply:
I get that from the beginning. That's why I think it's a very good catch! If you still think it's worth it after seeing the code, alright then~